## 0. 베이즈 정리

### 사건 E가 발생했다는 가정하에 사건 F가 발생할 확률이 주어짐. 그 때, 사건 F가 발생했다는 가정하에 사건 E가 발생할 확률은...? 
- 1) 조건부 확률 정의를 이용하여 식 정리
    - P(E|F) = P(E,F)/P(F) = P(F|E)P(E)/P(F)
- 2) 사건 F를 E의 발생 여부에 따라 분리
    - P(F) = P(F,E) + P(F,~E)
- 1)식에 2)식 대입하여 __베이즈 정리__ 완성!
    - P(E|F) = P(F|E)P(E) / [P(F|E)P(E) + P(F|~E)P(E)]

### 예제) 
- __전제__ : 10,000명 중 1명이 질병 d에 걸림. 질병여부를 판단하는 검사가 99%의 경우에 대해 정확한 판단을 내림.
- 양성 판정일 경우, 실제로 병에 걸렸을 확률 __P(D|T)__은??

## 1. 바보 스팸 필터

- S: 메시지가 스팸인 경우, V: '비아그라'라는 단어가 포함되어 있는 경우 라고 할때,
- 베이즈 정리를 사용하여 비아그라라는 단어가 포함되었을 때, 해당 메시지가 스팸일 확률:
    - P(S|V) = P(V|S)P(S) / [ P(V|S)P(S) + P(V|~S)P(~S) ] 
    - P(V|S)P(S) : 스팸이면서 비아그라 단어를 포함하고 있을 확률. -> __비아그라라는 단어가 포함된 모든 메시지 중 스팸 메시지의 비율.__
- 수많은 스팸 메시지와 스팸이 아닌 메시지를 갖고 있다면, P(V|S)와 P(V|~S)를 쉽게 추정할 수 있음.
- 메시지가 스팸일 확률과 스팸이 아닌 확률이 동일하다면(즉, P(S)=P(~S)=0.5) 식을 쉽게 정리.
    - P(S|V) = P(V|S) / [ P(V|S) + P(V|~S) ] 

### 예제)
- 스팸 메시지 중 50%, 스팸이 아닌 메시지 중 1%만이 비아그라라는 단어를 포함하고 있다면, 비아그라라는 단어를 포함하고 있는 메시지가 스팸일 확률은..?
- 0.5 / (0.5 + 0.01) = __98%__

## 2. 조금 더 똑똑한 스팸 필터
- 더 많은 단어 $w_1$ , ... $w_n$ 이 주어졌다고 가정.
- 확률 이론을 적용하기 위해 단어 $w_i$가 메시지에 포함되는 경우를 $X_i$로 나타내고, 스팸메시지에 i번째 단어가 포함되어 있는 확률인 P($X_i$|S)와 스팸이 아닌 메시지에 i번째 단어가 포함되어 있는 확률인 P($X_i$|~S)가 주어졌다고 표현.
- __나이브 베이즈의 핵심__ : 메시지가 스팸이냐 아니냐가 주어졌다는 조건 하에 각 단어의 존재(혹은 부재)는 서로 __조건부 독립적__이라는 가정에 기반.
    - 어떤 스팸 메시지가 '비아그라'라는 단어를 포함하고 있다는 점이 같은 메시지가 '롤렉스'라는 단어를 포함하고 있는지 판단하는데 도움을 주지 않음. (무관한 사건)

$$P(X_1=x_1, ...,X_n=x_n|S)=P(X_1=x_1|S) * ... * P(X_n=x_n|S)$$

- 비아그라라는 단어만으로 스팸을 걸러내는 필터에서도 사용된 베이즈 정리를 통해 메시지가 스팸일 확률을 다음과 같이 계산 할 수 있음.
    - P(S|X=x)=P(X=x|S)/ [ P(X=x|S) + P(X=x|~S) ] 
- 나이브 베이즈의 가정을 따르면, 각 단어가 메시지에 포함될 확률 값을 모두 곱해서 우변에 위치한 확률 값을 모두 구할 수 있음!
- __주의: 확률값을 실제로 곱하는 것을 피해야함!!__ -> 컴퓨터가 0에 가까운 부동소수점을 제대로 처리하지 못하여 언더플로 문제가 발생하기 때문. 
    - -> log의 성질을 이용하여 문제를 해결!!
        - log(ab) = log(a) + log(b)
        - exp(log x) = x
- P($X_i$|S),P($X_i$|~S) 추정 : 스팸으로 레이블된 학습 메시지 중 $w_i$가 포함되어 있는 메시지의 비율을 사용.
    - -> __단점:__ 주어진 학습 데이터에 '데이터'라는 단어는 스팸이 아닌 메시지에만 포함되어 있다고 상상해보자. 즉, P('데이터'|S) = 0 임. 그렇다면 나이브 베이즈 분류기는 '데이터'라는 단어가 들어간 메시지를 항상 스팸 메시지가 아니라고 예측할 것임. 메시지가 "값싼 비아그라와 정품 롤렉스에 대한 데이터"일대도 스팸이 아니라고 예측하게 됨.
    - -> __문제해결방법: Smoothing(평활화) 이용.__
- Smoothing을 위해 가짜 빈도수(pseudocount) k를 결정하고 스팸 메시지에서 i번째 단어가 나올 확률을 다음과 같이 추정.
    - P($X_i$|S) = (k + $w_i$를 포함하고 있는 스팸 수)/(2k+스팸 수)
    - -> 스팸 메시지에서 i번째 단어가 나올 확률을 계산할 때 해당 단어가 포함된 스팸과 포함되지 않은 스팸이 이미 각각 k번씩 나왔다고 가정한다는 의미. P('데이터'|S)를 0이 아닌 다른 확률 값으로 설정해줌.

## 3. 구현하기

In [1]:
from collections import Counter, defaultdict
# from machine_learning import split_data
import math, random, re, glob

### 형태소 분석기 정의

In [2]:
def tokenize(message):
    message = message.lower()                       # 소문자로 변환
    all_words = re.findall("[a-z0-9']+", message)   # 단어 추출
    return set(all_words)                           # 중복되는 단어 제거

In [3]:
sample_text = 'I have a pen. I have an apple. Uh! apple pen~'

In [4]:
tokenize(sample_text)

{'a', 'an', 'apple', 'have', 'i', 'pen', 'uh'}

### 빈도수 카운팅
- 각 단어를 key로 사용하며 __[스팸 메시지의 빈도수(spam_count), 스팸이 아닌 메시지의 빈도수(non_spam_count)]__형태의 list를 값으로 사용하는 dict를 반환해 줄 것임.

In [5]:
def count_words(training_set):
    """학습 데이터는 (message, is_spam) 형식으로 구성"""
    counts = defaultdict(lambda: [0,0])
    for message, is_spam in training_set:
        for word in tokenize(message):
            counts[word][0 if is_spam else 1] += 1
    return counts

### Smoothing
- 단어의 빈도수로 확률 값을 추정.

In [6]:
def word_probabilities(counts, total_spams, total_non_spams, k=0.5):
    """단어의 빈도수를 w, p(w | spam) and p(w | ~spam) 형태로 변환"""
    return [(w,
             (spam + k) / (total_spams + 2 * k),
             (non_spam + k) / (total_non_spams + 2 * k))
             for w, (spam, non_spam) in counts.items()]

### 스팸메시지 확률 계산
- __계산된 단어의 확률__과 __나이브 베이즈의 기본가정__을 사용해서 메시지가 스팸일 확률 계산.

In [7]:
def spam_probability(word_probs, message):
    message_words = tokenize(message) 
    log_prob_if_spam = log_prob_if_not_spam = 0.0

    # 모든 단어에 대해 반복.
    for word, prob_if_spam, prob_if_not_spam in word_probs:

        # 만약 메시지에 word가 나타나면 해당단어가 나올 log 확률 더해줌.
        if word in message_words:
            log_prob_if_spam += math.log(prob_if_spam)
            log_prob_if_not_spam += math.log(prob_if_not_spam)

        # 그렇지 않으면 해당 단어가 나오지 않을 log 확률 더해줌. 
        # 나오지 않을 확률은 log(1-나올 확률)로 계산
        else:
            log_prob_if_spam += math.log(1.0 - prob_if_spam)
            log_prob_if_not_spam += math.log(1.0 - prob_if_not_spam)

    prob_if_spam = math.exp(log_prob_if_spam)
    prob_if_not_spam = math.exp(log_prob_if_not_spam)
    return prob_if_spam / (prob_if_spam + prob_if_not_spam)

### 나이브 베이즈 클래스 정의

In [8]:
class NaiveBayesClassifier:

    def __init__(self, k=0.5):
        self.k = k
        self.word_probs = []

    def train(self, training_set):

        # 스팸과 아닌것의 개수 세어줌
        num_spams = len([is_spam
                         for message, is_spam in training_set
                         if is_spam])
        num_non_spams = len(training_set) - num_spams

        # run training data through our "pipeline"
        word_counts = count_words(training_set)
        self.word_probs = word_probabilities(word_counts,
                                             num_spams,
                                             num_non_spams,
                                             self.k)

    def classify(self, message):
        return spam_probability(self.word_probs, message)



## 4. 모델 검증하기
- 실습 데이터 : http://spamassassin.apache.org/publiccorpus/
    - spamc 라이브러리 이용해도 됨.

In [34]:
#!pip install spamc

In [18]:
# import spamc

SyntaxError: invalid syntax (client.py, line 230)

In [35]:
import glob, re, os

In [39]:
path = os.getcwd()+'/spam_data/*/*'
path

'/Users/chad/workspace/nlp_practice/naiveornot/spam_data/*/*'

In [38]:
data = []

In [45]:
for fn in glob.glob(path):
    is_spam = "ham" not in fn
    
    with open(fn, 'r',encoding='ISO-8859-1') as f:
        for line in f.readlines():
            if line.startswith("Subject:"):
                subject = re.sub(r"^Subject: ","", line).strip()
                data.append((subject, is_spam))

In [48]:
data[:30]

[('Re: New Sequences Window', False),
 ('[zzzzteana] RE: Alexander', False),
 ('[zzzzteana] Moscow bomber', False),
 ("[IRR] Klez: The Virus That  Won't Die", False),
 ('Re: Insert signature', False),
 ('Re: [zzzzteana] Nothing like mama used to make', False),
 ('Re: [zzzzteana] Nothing like mama used to make', False),
 ('[zzzzteana] Playboy wants to go out with a bang', False),
 ('Re: [zzzzteana] Nothing like mama used to make', False),
 ('[zzzzteana] Meaningful sentences', False),
 ('Re: New Sequences Window', False),
 ('[SAtalk] SA CGI Configurator Scripts', False),
 ('[SAdev] Interesting approach to Spam handling..', False),
 ('Re: New Sequences Window', False),
 ('[SAdev] Party in San Francisco tonight', False),
 ('Re: [SAdev] Live Rule Updates after Release ???', False),
 ('[ILUG] Re: Problems with RAID1 on cobalt raq3', False),
 ('Re: New Sequences Window', False),
 ('The case for spam', False),
 ("[IIU] Eircom aDSL Nat'ing", False),
 ('[zzzzteana] Re: Australian Catholic Kiddie

In [50]:
random.seed(0)

In [51]:
def split_data(data, prob):
    results = [], []
    for row in data:
        results[0 if random.random() < prob else 1].append(row)
    return results

In [52]:
train_data, test_data = split_data(data, 0.75)

In [54]:
classifier = NaiveBayesClassifier()

In [55]:
classifier.train(train_data)

In [60]:
# [제목, 실제 스팸 여부, 예측된 스팸 확률]
classified = [(subject, is_spam, classifier.classify(subject)) for subject, is_spam in test_data]

In [67]:
for case in classified[-100:]:
    print('{:20}\t{:>20}\t{:>20}'.format(case[0], case[1], case[2]))

(_) Increase Your Sales!!!!	                   1	  0.9571532542918919
Look What Sandy is Doing in her DORM!!	                   1	  0.9999956709881644
Attn: How about being a few pounds lighter?OLIBYKN	                   1	  0.9999989352520285
ADV: Extended Auto Warranties Here                                                   aanfh	                   1	  0.9999811983230319
The TBA Doctor Walks the Walk on Diabetes	                   1	  0.9999410847225025
How would you like to create your own DVD library? 4863d-5	                   1	  0.9998670234022997
Approved Online, Internet, Buy Viagra              SJHFL	                   1	   0.995177679738722
Testing a system, please delete	                   1	 0.00269205388369314
ADV:  Harvest lots of E-mail addresses quickly !	                   1	  0.9119642164250001
Place Your LTC Declines with Us	                   1	 0.01641826015932076
Welcome to the concept of RESIDUAL Income!	                   1	0.012079883329474326
URGENT BUSINESS

In [57]:
# 메시지가 스팸일 확류이 0.5 이상이면 스팸이라고 하자. 예측된 스팸 메시지가 실제 스팸인 경우를 세어보면..
counts = Counter((is_spam, spam_probability> 0.5) for _, is_spam, spam_probability in classified)

In [68]:
counts

Counter({(False, False): 704,
         (False, True): 33,
         (True, False): 38,
         (True, True): 101})

- 실제 스팸 -> 스팸 예측(TP) : 101
- 실제 햄 -> 스팸 예측(FP) : 33
- 실제 햄 -> 햄 예측(TN) : 704
- 실제 스팸 -> 햄 예측(FN) : 38

- 정밀도(Precision) : 101/(101+33) = 75%
- 재현율(Recall) : 101/(101+38) = 73%

In [75]:
# 스팸일 확률 오른차순으로 정렬:
classified.sort(key=lambda row: row[2])

In [78]:
for case in classified[-10:]:
    print('{:20}\t{:>20}\t{:>20}'.format(case[0], case[1], case[2]))

One of a kind Money maker! Try it for free!	                   1	  0.9999998550195005
ADV: Interest rates slashed! Don't wait!                                                   ccaxc	                   1	    0.99999987256979
Adv: Mortgage Quotes Fast Online, No Cost	                   1	  0.9999999135269964
ADV: Are you protected? Free Life Insurance quote oefik	                   1	  0.9999999707931284
FORTUNE 500 WORK AT HOME REPS NEEDED!	                   1	   0.999999981456028
FORTUNE 500 WORK AT HOME REPS NEEDED!	                   1	   0.999999981456028
Attn: NORTON SYSTEMWORKS 2002 CLEARANCE SALE!              VOJTGT	                   1	  0.9999999883165344
HOME REPS WANTED-FORTUNE 500 COMP HIRING	                   1	  0.9999999906606079
MSNBC: Rates Hit 18 year Low 4.75% ...28940	                   1	    0.99999999435782
Online Doctors will fill your Viagra Prescription Now!!!                QEEB	                   1	  0.9999999985329022


In [ ]:
# 스팸이 아닌 메시지 중 스팸일 확률이 가장 높은 메시지

In [ ]:
spammiest_hams = filter(lambda row:)

In [20]:
def get_subject_data(path):

    data = []

    # regex for stripping out the leading "Subject:" and any spaces after it
    subject_regex = re.compile(r"^Subject:\s+")

    # glob.glob returns every filename that matches the wildcarded path
    for fn in glob.glob(path):
        is_spam = "ham" not in fn

        with open(fn,'r',encoding='ISO-8859-1') as file:
            for line in file:
                if line.startswith("Subject:"):
                    subject = subject_regex.sub("", line).strip()
                    data.append((subject, is_spam))

    return data


In [21]:
def p_spam_given_word(word_prob):
    word, prob_if_spam, prob_if_not_spam = word_prob
    return prob_if_spam / (prob_if_spam + prob_if_not_spam)


In [22]:
def train_and_test_model(path):

    data = get_subject_data(path)
    random.seed(0)      # just so you get the same answers as me
    train_data, test_data = split_data(data, 0.75)

    classifier = NaiveBayesClassifier()
    classifier.train(train_data)

    classified = [(subject, is_spam, classifier.classify(subject))
              for subject, is_spam in test_data]

    counts = Counter((is_spam, spam_probability > 0.5) # (actual, predicted)
                     for _, is_spam, spam_probability in classified)

    print(counts)

    classified.sort(key=lambda row: row[2])
    spammiest_hams = list(filter(lambda row: not row[1], classified))[-5:]
    hammiest_spams = list(filter(lambda row: row[1], classified))[:5]

    print("spammiest_hams", spammiest_hams)
    print("hammiest_spams", hammiest_spams)

    words = sorted(classifier.word_probs, key=p_spam_given_word)

    spammiest_words = words[-5:]
    hammiest_words = words[:5]

    print("spammiest_words", spammiest_words)
    print("hammiest_words", hammiest_words)

In [23]:
if __name__ == "__main__":
    #train_and_test_model(r"c:\spam\*\*")
    train_and_test_model(r"/home/joel/src/spam/*/*")

NameError: name 'split_data' is not defined